solc/CommandLineInterface.cpp

```diff
@@ -990,10 +990,15 @@ void CommandLineInterface::handleAst(string const& _argStr)
 		map<ASTNode const*, eth::GasMeter::GasConsumption> gasCosts;
 		// FIXME: shouldn't this be done for every contract?
 		if (m_compiler->runtimeAssemblyItems(m_compiler->lastContractName()))
-			gasCosts = GasEstimator::breakToStatementLevel(
+		{
+			//NOTE: keep the local variable `ret` to prevent a Heisenbug that could happen on certain mac os platform.
+			//See: https://github.com/ethereum/solidity/issues/3718 for details.
+			auto ret = GasEstimator::breakToStatementLevel(
 				GasEstimator(m_evmVersion).structuralEstimation(*m_compiler->runtimeAssemblyItems(m_compiler->lastContractName()), asts),
 				asts
 			);
+			gasCosts = ret;
```

Comment:
Could you check if the bug is present when we use `gasCosts = std::move(ret);` here?

Reply:
After I added `gasCosts = std::move(ret);` the bug is still present.

Reply:
Mac OS: 
```
El Capitan 10.11.6 (15G21013)
```

Xcode: 
not used

boost:
```sh
>> grep "#define BOOST_VERSION" /usr/local/include/boost/version.hpp
#define BOOST_VERSION_HPP
#define BOOST_VERSION 106600
```

clang: 
```
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

Reply:
Thanks a lot (and also thanks for your patience)!
One important piece of information is also the version of the standard library you link against - is that directly visible from the clang version or the target?

Reply:
updated above

Reply:
@alibabe @chriseth The bug indeed doesn't seem to be present when linking newer versions of the standard library. I can't reproduce it on my machine with the following setup: 

macOS Sierra 10.12.6
Clang 9.0.0

and ``otool -L solc`` outputs:
```
/usr/local/opt/z3/lib/libz3.dylib (compatibility version 0.0.0, current version 0.0.0)
/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 307.5.0)
/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
```

I think we should try to selectively apply this workaround to older versions of macOS / clang since it can be really annoying to people that don't want or can't update their system.

Reply:
@alibabe Could you please check the libc++ version number that is defined through ``_LIBCPP_VERSION`` in ``/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__config``?

Reply:
@bit-shift I have `#define _LIBCPP_VERSION 3700` on my environment

Reply:
Great! Thanks. That means we could conditionally compile the workaround using (haven't tested with clang on Linux and Windows though)
```
#if defined(_LIBCPP_VERSION) && (_LIBCPP_VERSION == 3700)
	#define LIBCPP_MOVE_WORKAROUND
#endif
```
@chriseth What do you think? I'm not sure if we should fix it to this one version or also apply it to older ones. Perhaps even ``_LIBCPP_VERSION < 4000``?

Reply:
I would use `_LIBCPP_VERSION < 4000`, because we do not know when it was fixed.

Reply:
@bit-shift How do you its the libc++ that's causing the issue? 
@chriseth even if we do `_LIBCPP_VERSION < 4000` it's not guaranteed to fix the issue, right? because we don't know when it was fixed